lib/temporal/data_converter.rb

```diff
@@ -1,40 +1,73 @@
+require 'temporal/api/common/v1/message_pb'
+require 'temporal/errors'
 require 'temporal/payload_converter'
 
 module Temporal
   class DataConverter
-    def initialize(payload_converter: Temporal::PayloadConverter::DEFAULT)
+    class MissingPayload < Temporal::Error; end
+
+    def initialize(payload_converter: Temporal::PayloadConverter::DEFAULT, payload_codecs: [])
```

Comment:
You might want to instead take a class name for the `payload_converter` here and instantiate it reflectively if at all possible. We learned this a bit late in Python, but when we build the sandbox, we want the converter to run _inside_ the sandbox which means we need to instantiate it there.

At the very least accepting a payload converter class may help users understand that payload converters have to be deterministic too. Note, this doesn't apply to codecs. While data converters combine the two (and is useful for some client/activity things), the way workflows work you'll want to apply the codec across entire activations/commands and use the payload converter with the rest of the workflow code instead of using the data converter directly.

Sorry if this doesn't make sense.

Reply:
I mean you can already submit a class if you want as long as it implements the expected interface. As for instantiating the converter ourselves — you'd need to submit all the input args to it, which makes it difficult with the `Composite` converter that itself expects encoding converters that might need to be initialized as well.

> While data converters combine the two (and is useful for some client/activity things), the way workflows work you'll want to apply the codec across entire activations/commands and use the payload converter with the rest of the workflow code instead of using the data converter directly.

I'm happy to change this as we get closer to Workflow implementation. Right now unfortunately it doesn't make much sense because I'm not 100% sure what that interaction will look like.

Reply:
:+1: Can wait on this. In Python we have a `DefaultPayloadConverter` class that just extends `CompositePayloadConverter` so that just the class name can be passed around.

Reply:
Let's wait then 👍 

> In Python we have a DefaultPayloadConverter class that just extends CompositePayloadConverter so that just the class name can be passed around

Does that mean that your `PayloadConverter` interface now also includes a constructor that takes no arguments?

Reply:
Yes, and I require such a thing because I must be able to instantiate it in the sandbox. No state is shared. The same thing for workflow interceptors. _technically_ I could abstract further and take a factory or accept serializable args to create on the other side of the sandbox boundary, but I'd rather just let the user do that.